twenty-first/src/shared_math/polynomial.rs

```diff
@@ -197,135 +135,67 @@ impl<PFElem: FiniteField> Polynomial<PFElem> {
-        p1: (PFElem, PFElem),
-        p2: (PFElem, PFElem),
-    ) -> bool {
-        if p0.0 == p1.0 || p1.0 == p2.0 || p2.0 == p0.0 {
-            return false;
+        if degree == 0 {
+            return Self::from_constant(self.coefficients[0] * self.coefficients[0]);
         }
 
-        let dy = p0.1 - p1.1;
-        let dx = p0.0 - p1.0;
+        let result_degree: u64 = 2 * self.degree() as u64;
+        let order = roundup_npo2(result_degree + 1);
+        let root_res = BFieldElement::primitive_root_of_unity(order);
+        let root = match root_res {
+            Some(n) => n,
+            None => panic!("Failed to find primitive root for order = {}", order),
+        };
 
-        dx * (p2.1 - p0.1) == dy * (p2.0 - p0.0)
-    }
+        let mut coefficients = self.coefficients.to_vec();
+        coefficients.resize(order as usize, PFElem::zero());
+        let log_2_of_n = log_2_floor(coefficients.len() as u128) as u32;
+        ntt::<PFElem>(&mut coefficients, root, log_2_of_n);
 
-    pub fn get_colinear_y(p0: (PFElem, PFElem), p1: (PFElem, PFElem), p2_x: PFElem) -> PFElem {
-        debug_assert_ne!(p0.0, p1.0, "Line must not be parallel to y-axis");
-        let dy = p0.1 - p1.1;
-        let dx = p0.0 - p1.0;
-        let p2_y_times_dx = dy * (p2_x - p0.0) + dx * p0.1;
+        for element in coefficients.iter_mut() {
+            *element = element.to_owned() * element.to_owned();
+        }
 
-        // Can we implement this without division?
-        p2_y_times_dx / dx
-    }
+        intt::<PFElem>(&mut coefficients, root, log_2_of_n);
+        coefficients.truncate(result_degree as usize + 1);
 
-    pub fn zerofier(domain: &[PFElem]) -> Self {
-        if domain.is_empty() {
-            return Self {
-                coefficients: vec![PFElem::one()],
-            };
-        }
-        let mut zerofier_array = vec![PFElem::zero(); domain.len() + 1];
-        zerofier_array[0] = PFElem::one();
-        let mut num_coeffs = 1;
-        for &d in domain.iter() {
-            for k in (1..num_coeffs + 1).rev() {
-                zerofier_array[k] = zerofier_array[k - 1] - d * zerofier_array[k];
-            }
-            zerofier_array[0] = -d * zerofier_array[0];
-            num_coeffs += 1;
-        }
-        Self {
-            coefficients: zerofier_array,
-        }
+        Polynomial { coefficients }
     }
 
-    // Slow square implementation that does not use NTT
     #[must_use]
-    pub fn slow_square(&self) -> Self {
+    pub fn square(&self) -> Self {
         let degree = self.degree();
         if degree == -1 {
             return Self::zero();
         }
 
+        // A benchmark run on sword_smith's PC revealed that
+        // `fast_square` was faster when the input size exceeds
+        // a length of 64.
         let squared_coefficient_len = self.degree() as usize * 2 + 1;
+        if squared_coefficient_len > 64 {
+            return self.fast_square();
+        }
```

Comment:
Scientists say... ;-)

Maybe we can do a benchmark here.

This seems like a thing that could be quite machine-dependent.

Reply:
One data point is better than zero.